gnomad_qc/v3/annotations/generate_freq_data.py

```diff
@@ -1,104 +1,260 @@
-    )
+            logger.info(f"Writing out frequency data for {subset} subset...")
+            if args.test:
+                mt.rows().write(
+                    f"gs://gnomad-tmp/gnomad_freq/chr20_test_freq.{subset}.ht",
+                    overwrite=True,
+                )
+            else:
+                mt.rows().write(get_freq(subset=subset).path, overwrite=args.overwrite)
 
-    logger.info("Writing out frequency data...")
-    if args.test:
-        mt.rows().write("gs://gnomad-tmp/gnomad_freq/chr20_1_1000000_freq.ht", overwrite=True)
-    else:
-        mt.rows().write(freq.path, overwrite=args.overwrite)
+        else:
+            logger.info("Computing age histograms for each variant...")
+            mt = mt.annotate_rows(
+                **age_hists_expr(
+                    mt.adj,
+                    mt.GT,
+                    hl.if_else(
+                        hl.is_defined(mt.meta.project_meta.age),
+                        mt.meta.project_meta.age,
+                        mt.meta.project_meta.age_alt,
+                    ),
+                )
+            )
+
+            # Compute callset-wide age histogram global
+            meta_ht = meta.ht()
+            meta_ht = meta_ht.filter(meta_ht.release)
+            mt = mt.annotate_globals(
+                age_distribution=meta_ht.aggregate(
+                    hl.agg.hist(
+                        hl.if_else(
+                            hl.is_defined(meta_ht.project_meta.age),
+                            meta_ht.project_meta.age,
+                            meta_ht.project_meta.age_alt,
+                        ),
+                        30,
+                        80,
+                        10,
+                    )
+                )
+            )
+
+            mt = annotate_freq(
+                mt,
+                sex_expr=mt.meta.sex_imputation.sex_karyotype,
+                pop_expr=mt.meta.population_inference.pop,
+                downsamplings=DOWNSAMPLINGS,
+            )
+            # Remove all loci with raw AC=0
+            mt = mt.filter_rows(mt.freq[1].AC > 0, keep=True)
+
+            logger.info("Calculating InbreedingCoefficient...")
+            # NOTE: This is not the ideal location to calculate this, but added here to avoid another densify
+            mt = mt.annotate_rows(
+                InbreedingCoeff=bi_allelic_site_inbreeding_expr(mt.GT)
+            )
+
+            # Select freq, FAF and popmax
+            faf, faf_meta = faf_expr(
+                mt.freq, mt.freq_meta, mt.locus, POPS_TO_REMOVE_FOR_POPMAX
+            )
+            mt = mt.select_rows(
+                "InbreedingCoeff",
+                "freq",
+                faf=faf,
+                popmax=pop_max_expr(mt.freq, mt.freq_meta, POPS_TO_REMOVE_FOR_POPMAX),
+            )
+            mt = mt.annotate_globals(faf_meta=faf_meta)
+            mt = mt.annotate_globals(faf_index_dict=make_faf_index_dict(mt))
```

Comment:
I thought that `faf_expr` returned a list because it returns `hl.eval(faf_meta)`. I think if you use just `faf_meta` not `mt.faf_meta` this should still work, and then you don't need to do the `hl.eval` because that is already how it was returned by `faf_expr`

Reply:
Oh, I see what you're doing. I read the final line as `make_faf_index_dict(mt.faf_meta)` (referencing the new mt global created in the previou line), but you're just passing in the result from `faf_meta` again. I think this should work then. Will change to your suggestion

Reply:
Confirmed in a small test:
```
from typing import List, Dict, Tuple

import hail as hl

from gnomad.resources.grch38.gnomad import (
    DOWNSAMPLINGS,
    POPS,
    POPS_TO_REMOVE_FOR_POPMAX,
    SEXES,
)
from gnomad.utils.annotations import (
    annotate_freq,
    faf_expr,
    get_adj_expr,
)
from gnomad.utils.vcf import index_globals
from gnomad_qc.v3.resources.annotations import get_freq
from gnomad_qc.v3.resources.basics import get_gnomad_v3_mt


def make_faf_index_dict(faf_meta: List[Dict[str, str]]):
    """
    Create a look-up Dictionary for entries contained in the filter allele frequency annotation array
This conversation was marked as resolved by gtiao
    :param List of Dict faf_meta: Global annotation containing the set of groupings for each element of the faf array (e.g., [{'group': 'adj'}, {'group': 'adj', 'pop': 'nfe'}])
    :return: Dictionary of faf annotation population groupings, where values are the corresponding 0-based indices for the
        groupings in the faf_meta array
    :rtype: Dict of str: int
    """

    index_dict = index_globals(faf_meta, dict(group=["adj"]))
    index_dict.update(index_globals(faf_meta, dict(group=["adj"], pop=POPS)))
    index_dict.update(index_globals(faf_meta, dict(group=["adj"], sex=SEXES)))
    index_dict.update(index_globals(faf_meta, dict(group=["adj"], pop=POPS, sex=SEXES)))

    return index_dict


mt = get_gnomad_v3_mt(
            key_by_locus_and_alleles=True, release_only=True, samples_meta=True
        )
mt = mt._filter_partitions(range(2))
mt = hl.experimental.sparse_split_multi(mt, filter_changed_loci=True)
mt = hl.experimental.densify(mt)
mt = mt.annotate_entries(
        adj=get_adj_expr(mt.GT, mt.GQ, mt.DP, mt.AD),
    )
mt = mt.filter_rows(hl.len(mt.alleles) > 1)
mt = annotate_freq(
                mt,
                sex_expr=mt.meta.sex_imputation.sex_karyotype,
                pop_expr=mt.meta.population_inference.pop,
                downsamplings=DOWNSAMPLINGS,
            )
faf, faf_meta = faf_expr(
                mt.freq, mt.freq_meta, mt.locus, POPS_TO_REMOVE_FOR_POPMAX
            )
mt = mt.annotate_globals(faf_meta=faf_meta)
mt = mt.annotate_globals(
    faf_meta=faf_meta,
    faf_index_dict=make_faf_index_dict(faf_meta)
)
print(mt.faf_index_dict.collect())
```
returns [{'amr-XY-adj': 17, 'sas-XY-adj': 16, 'nfe-XX-adj': 8, 'sas-adj': 4, 'amr-XX-adj': 12, 'XX-adj': 6, 'sas-XX-adj': 11, 'XY-adj': 7, 'eas-adj': 3, 'amr-adj': 5, 'afr-adj': 2, 'eas-XY-adj': 15, 'adj': 0, 'afr-XX-adj': 9, 'afr-XY-adj': 14, 'eas-XX-adj': 10, 'nfe-XY-adj': 13, 'nfe-adj': 1}]